### ISSUE TITLE
Package signature verification is slow on Linux

### ISSUE THREAD

**Issue by marcin-krystianc** [NuGet/Home#12672]
### NuGet Product Used

dotnet.exe

### Product Version

7.0.107, 8.0.100-preview.6.23319.41

### Impact

It bothers me. A fix would be nice

### Repro Steps & Context

https://github.com/marcin-krystianc/TestNuGetSignatureVerification
### TestNuGetSignatureVerification on Linux
``` <docker.exe run -t -i --rm ubuntu>
> apt-get update && apt-get install -y git dotnet7
> dotnet --info
> git clone https://github.com/marcin-krystianc/TestNuGetSignatureVerification.git
> git clone https://github.com/dotnet/orleans.git
> dotnet restore orleans/Orleans.sln
> export DOTNET_NUGET_SIGNATURE_VERIFICATION=True
> dotnet run -c Release --project TestNuGetSignatureVerification/TestNuGetSignatureVerification/TestNuGetSignatureVerification.csproj -- 1
Using MSBuild:/usr/lib/dotnet/sdk/7.0.107
Found 479 packages in '/root/.nuget/packages/
Verified 479 packages in '31.7100083 seconds with degree of parallelism=1

> dotnet nuget locals --list global-packages
> time dotnet nuget verify --all ~/.nuget/packages/**/*/*.nupkg -v diag > out.txt
real    0m47.733s
user    0m48.614s
sys     0m0.983s
```

### TestNuGetSignatureVerification on Windows (PowerShell)
```
> dotnet --info
> git clone https://github.com/marcin-krystianc/TestNuGetSignatureVerification.git
> git clone https://github.com/dotnet/orleans.git
> dotnet restore orleans/Orleans.sln
> dotnet run -c Release --project TestNuGetSignatureVerification/TestNuGetSignatureVerification/TestNuGetSignatureVerification.csproj -- 1
Using MSBuild:C:\Program Files\dotnet\sdk\7.0.304
Found 493 packages in 'C:\Users\Marcin\.nuget\packages\
Verified 493 packages in '4.1218048 seconds with degree of parallelism=1

> dotnet nuget locals --list global-packages
> Measure-Command {dotnet nuget verify --all $env:USERPROFILE\.nuget\packages\**\*\*.nupkg -v diag > out.txt}
...
Seconds           : 4
Milliseconds      : 638
...
```

I would expect that signature verification on Linux and on Windows takes a similar time. Unfortunately, signature verification on Linux is about an order of magnitude slower. It affects the "clean" restore scenarios, where the NuGet packages need to be installed into `global-packages` folder as the signature verification was recently enabled by default on Linux.

It is probably not even a problem with NuGet itself but with how the signature verification is implemented on Windows and on Linux. I'm reporting it here as I've discovered it with NuGet and also it provides an easy way to reproduce it.

**Comment by nkolev92** [comment]
479 vs 493 package probably makes the comparison not 100% equal, but the difference shouldn't be 10x. 

cc @dtivel are you familiar with any issues here?

**Comment by erdembayar** [comment]
Thank you for reporting this issue. Package signature validation depends on many ambient factors, such as the certificate revocation list. Could you please try turning off the validation by using an environmental variable and then try again?

Happy coding! 🧑‍💻

**Comment by marcin-krystianc** [comment]
Hi @erdembayar ,

I've tried different values for `NUGET_CERT_REVOCATION_MODE` (`Offline` and `Online`) but it didn't make any significant difference.
Regarding the `DOTNET_NUGET_SIGNATURE_VERIFICATION`, it really affects the restore operation. (PS. I'm using `--disable-parallel` because the results are less noise than without it.)
```
root@4db7832c41c4:/orleans# dotnet --version
7.0.305
root@4db7832c41c4:/orleans# for i in {1..10};do dotnet nuget locals -c global-packages && DOTNET_NUGET_SIGNATURE_VERIFICATION=False dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true --disable-parallel; done | grep -i Elapsed
Time Elapsed 00:00:14.57
Time Elapsed 00:00:10.55
Time Elapsed 00:00:12.46
Time Elapsed 00:00:11.85
Time Elapsed 00:00:12.03
Time Elapsed 00:00:10.02
Time Elapsed 00:00:09.83
Time Elapsed 00:00:11.21
Time Elapsed 00:00:09.76
Time Elapsed 00:00:18.09
root@4db7832c41c4:/orleans# for i in {1..10};do dotnet nuget locals -c global-packages && DOTNET_NUGET_SIGNATURE_VERIFICATION=True dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true --disable-parallel; done | grep -i Elapsed
Time Elapsed 00:00:35.81
Time Elapsed 00:00:29.04
Time Elapsed 00:00:29.67
Time Elapsed 00:00:32.48
Time Elapsed 00:00:29.65
Time Elapsed 00:00:33.99
Time Elapsed 00:00:29.56
Time Elapsed 00:00:28.82
Time Elapsed 00:00:28.37
Time Elapsed 00:00:31.43
```

**Comment by dtivel** [comment]
First, thank you, @marcin-krystianc, for the report and the excellent repro steps.  I was able to reproduce the behavior you describe.

I get 475 packages.  I don't think the difference (475 vs. 479 vs. 493) is significant.

All 475 packages are from https://nuget.org and have a repository signature and timestamp.  Looking at all 475 repository signature timestamps, there are only 3 distinct timestamp certificate chains.  NuGet client builds a certificate chain every time it validates a timestamp, so that's 475 chain builds for only 3 distinct certificate chains.  The same idea applies to repository signatures; there are 475 chain builds for only 2 distinct certificate chains.  To a much lesser extent, this phenomenon also applies to author signatures and their timestamps.  As an experiment, I added in NuGet caching of successful chain building results, such that chain building was performed only once per leaf certificate.  The overall time dropped by ~24%.  It's obvious from experimentation that Windows has some additional per-process caching which memoizes certificate chain building results, and we get this optimization for free on the Windows platform.  The Linux platform doesn't provide an equivalent optimization, so we'll have to look into if/where a similar optimization makes sense on Linux.

That experiment accounted for ~24% of the overall time.  As for the rest of the difference between Windows and Linux, I don't have clear culprits yet.  One trace suggested that ~33% of the overall time was spent in regular expression classes, but I can't corroborate this.

.NET already provides some caching of intermediate chain building results (CRL's, OCSP responses, and certificates fetched via AIA), which is great.  This is because the platform (Linux) doesn't provide equivalent behaviors by default.  I agree with your sentiment that this is "probably not even a problem with NuGet itself but with how the signature verification is implemented on Windows and on Linux."

For now, if the performance hit is unacceptable, you have the option of disabling verification using the [`DOTNET_NUGET_SIGNATURE_VERIFICATION` environment variable](https://learn.microsoft.com/en-us/dotnet/core/tools/nuget-signed-package-verification#linux).

**Comment by marcin-krystianc** [comment]
@dtivel Thanks for your detailed response. I think, that another factor for slow certificate verification on Linux is OpenSSL itself. There are one or more performance regressions in it (e.g. https://github.com/openssl/openssl/issues/20051). I've confirmed it by running my tests on an older version of Ubuntu, and the results were much better. I'll keep digging and will post a detailed response once I've precise data.

PS:
Here are some test results:
- Test script:
```
export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 \
&& apt-get update && apt-get install -y git wget \
&& mkdir dotnet \
&& wget https://download.visualstudio.microsoft.com/download/pr/87a55ae3-917d-449e-a4e8-776f82976e91/03380e598c326c2f9465d262c6a88c45/dotnet-sdk-7.0.305-linux-x64.tar.gz \
&& tar -C dotnet -xvzf dotnet-sdk-7.0.305-linux-x64.tar.gz \
&& rm dotnet-sdk-7.0.305-linux-x64.tar.gz \
&& echo "export DOTNET_ROOT=/dotnet" >> $HOME/.bashrc \
&& echo "export PATH=$PATH:/dotnet:$HOME/.dotnet/tools" >> $HOME/.bashrc \
&& source ~/.bashrc \
&& dotnet --info \
&& git clone https://github.com/dotnet/orleans.git \
&& export DOTNET_NUGET_SIGNATURE_VERIFICATION=False \
&& echo DOTNET_NUGET_SIGNATURE_VERIFICATION=False \
&& for i in {1..5};do dotnet nuget locals -c global-packages && dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true orleans/Orleans.sln --disable-parallel; done | grep -i Elapsed \
&& export DOTNET_NUGET_SIGNATURE_VERIFICATION=True \
&& echo DOTNET_NUGET_SIGNATURE_VERIFICATION=True \
&& for i in {1..5};do dotnet nuget locals -c global-packages && dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true orleans/Orleans.sln --disable-parallel; done | grep -i Elapsed
```

- ubuntu:18.04 (`openssl version` -> `OpenSSL 1.1.1  11 Sep 2018`):
```
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:01:26.31
Time Elapsed 00:00:09.75
Time Elapsed 00:00:09.45
Time Elapsed 00:00:09.45
Time Elapsed 00:00:09.59
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:13.39
Time Elapsed 00:00:13.00
Time Elapsed 00:00:12.65
Time Elapsed 00:00:12.58
Time Elapsed 00:00:12.56
```

- ubuntu:22.04 (`openssl version`  -> `OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)`):
```
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:01:14.51
Time Elapsed 00:00:09.76
Time Elapsed 00:00:09.64
Time Elapsed 00:00:09.51
Time Elapsed 00:00:09.65
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:28.77
Time Elapsed 00:00:28.37
Time Elapsed 00:00:28.32
Time Elapsed 00:00:28.47
Time Elapsed 00:00:28.89
```

**Comment by dtivel** [comment]
@marcin-krystianc, thank you for the additional information.  I tested on Debian 12 "Bookworm".  `openssl version` prints `OpenSSL 3.0.9 30 May 2023 (Library: OpenSSL 3.0.9 30 May 2023)`.

**Comment by adrien-n** [comment]
Hi, I am finishing preparing a SRU (stable release update) for openssl in Ubuntu 22.04. It includes a series of patch that improves performance and could help your usecase.

There is a PPA at https://launchpad.net/~adrien-n/+archive/ubuntu/openssl-jammy-sru which you can test and I would welcome performance numbers if you do so.

**Comment by marcin-krystianc** [comment]
Hi @adrien-n,

I've run another round of benchmarks. I've followed https://launchpad.net/~adrien-n/+archive/ubuntu/openssl-jammy-sru, but it didn't make a massive difference (only a few percent of improvement). Performance in Ubuntu 24.04 (OpenSSL 3.0.10 1 Aug 2023) is visibly faster, but still far from Ubuntu 20.04 (OpenSSL 1.1.1f  31 Mar 2020).

### Updated test script:
```
export DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 \
&& apt-get update && apt-get install -y git wget \
&& mkdir dotnet \
&& wget https://download.visualstudio.microsoft.com/download/pr/5226a5fa-8c0b-474f-b79a-8984ad7c5beb/3113ccbf789c9fd29972835f0f334b7a/dotnet-sdk-8.0.100-linux-x64.tar.gz \
&& tar -C dotnet -xvzf dotnet-sdk-8.0.100-linux-x64.tar.gz \
&& rm dotnet-sdk-8.0.100-linux-x64.tar.gz \
&& echo "export DOTNET_ROOT=/dotnet" >> $HOME/.bashrc \
&& echo "export PATH=$PATH:/dotnet:$HOME/.dotnet/tools" >> $HOME/.bashrc \
&& source ~/.bashrc \
&& git clone https://github.com/dotnet/orleans.git \
&& git --git-dir=/orleans/.git checkout 2e454ae58197ade4855392bdeffaa86bf3b588fd \
&& dotnet --info \
&& openssl version \
&& export DOTNET_NUGET_SIGNATURE_VERIFICATION=False \
&& echo DOTNET_NUGET_SIGNATURE_VERIFICATION=False \
&& for i in {1..5};do dotnet nuget locals -c global-packages && dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true orleans/Orleans.sln --disable-parallel; done | grep -i Elapsed \
&& export DOTNET_NUGET_SIGNATURE_VERIFICATION=True \
&& echo DOTNET_NUGET_SIGNATURE_VERIFICATION=True \
&& for i in {1..5};do dotnet nuget locals -c global-packages && dotnet restore -clp:summary /p:RestoreUseStaticGraphEvaluation=true orleans/Orleans.sln --disable-parallel; done | grep -i Elapsed
```

###  Results:
```
# ubuntu:22.04
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:01:26.04
Time Elapsed 00:00:09.83
Time Elapsed 00:00:09.46
Time Elapsed 00:00:09.53
Time Elapsed 00:00:09.66
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:30.28
Time Elapsed 00:00:29.51
Time Elapsed 00:00:29.30
Time Elapsed 00:00:29.42
Time Elapsed 00:00:29.69

# ubuntu:22.04 + https://launchpad.net/~adrien-n/+archive/ubuntu/openssl-jammy-sru:
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:00:09.90
Time Elapsed 00:00:09.28
Time Elapsed 00:00:10.51
Time Elapsed 00:00:10.08
Time Elapsed 00:00:10.55
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:29.47
Time Elapsed 00:00:28.36
Time Elapsed 00:00:28.50
Time Elapsed 00:00:28.46
Time Elapsed 00:00:28.67

# ubuntu:24.04
OpenSSL 3.0.10 1 Aug 2023 (Library: OpenSSL 3.0.10 1 Aug 2023)
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:01:27.11
Time Elapsed 00:00:09.55
Time Elapsed 00:00:09.12
Time Elapsed 00:00:10.31
Time Elapsed 00:00:09.44
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:26.43
Time Elapsed 00:00:25.42
Time Elapsed 00:00:24.93
Time Elapsed 00:00:25.23
Time Elapsed 00:00:24.64

# ubuntu:20.04
OpenSSL 1.1.1f  31 Mar 2020
DOTNET_NUGET_SIGNATURE_VERIFICATION=False
Time Elapsed 00:01:56.35
Time Elapsed 00:00:10.94
Time Elapsed 00:00:09.55
Time Elapsed 00:00:09.43
Time Elapsed 00:00:09.68
DOTNET_NUGET_SIGNATURE_VERIFICATION=True
Time Elapsed 00:00:13.84
Time Elapsed 00:00:12.48
Time Elapsed 00:00:13.10
Time Elapsed 00:00:12.72
Time Elapsed 00:00:13.92
```

**Comment by adrien-n** [comment]
Thanks a lot for testing. This seems to validate the patch is useful (if a single patch series brings 20% of the improvement of 1000 patch series, it seems interesting to me).

Unfortunately the performance is still way worse. It's possible that openssl 3.1 or 3.2 would be much faster. The issue with openssl for Ubuntu is that there is no more recent openssl LTS version and we want openssl LTS for our own LTS. Since there is also no calendar, this also prevents us from moving forward in interim releases since we don't know if there will be an LTS in time for our next LTS.

**Comment by saghm** [comment]
Is there anything that's being done for this issue? In trying to build something recently on Arch Linux (with OpenSSL 3.6.1, so much newer than anything references in this thread), trying to run `dotnet build` for a fresh clone of a project that's in the process of adding Linux support was taking over 2 minutes for a restore (which I didn't bother waiting to let finish), and running with `DOTNET_NUGET_SIGNATURE_VERIFICATION=false` seems to cause it to finish within 10 seconds or so. Seeing that this has been an open issue for nearly three years without any updates in over two years doesn't give me a lot of confidence in terms of how serious Linux support is for .NET.

**Comment by zivkan** [comment]
@saghm if you instead try setting `NUGET_CERT_REVOCATION_MODE` to `offline`, as has been recommended earlier in this thread, does that also resolve your issue?  This thread has been discussing openssl causing restore to take 25 seconds instead of 10 seconds, which is a completely different order of magnitude to the 2 minutes vs 10 seconds you're reporting.

NuGet uses X.509 certificates, just like web browsers https, to sign packages. X.509 certificates embed URLs so clients can check if the certificate has been revoked. If your network blocks those requests (especially if it drops the request without sending a TCP RST packet), then clients will wait for a timeout period. This is not a Linux issue, it'll happen on any OS. I recently started using Arch Linux myself, and haven't had any problems with minute long restores.

**Comment by saghm** [comment]
>  This thread has been discussing openssl causing restore to take 25 seconds instead of 10 seconds, which is a completely different order of magnitude to the 2 minutes vs 10 seconds you're reporting.

Understood. I saw that most of the prior discussion has been around openssl, but I wasn't sure whether it made more sense to open a separate issue given that at least the title seemed to be generic enough to apply to what I'm dealing with. I'm happy to open a separate issue if that's preferable though rather than diluting the discussion here.

> NuGet uses X.509 certificates, just like web browsers https, to sign packages. X.509 certificates embed URLs so clients can check if the certificate has been revoked. If your network blocks those requests (especially if it drops the request without sending a TCP RST packet), then clients will wait for a timeout period.

From my admittedly limited knowledge in this area, I'm not aware of any other package managers that use X.509. The only places I've seen them used in the past are for HTTPS like you mention and in connections to things like SSH servers, databases, and other similar services. The main common element of all of these from what I can tell is that the number of certificates; I'm not likely to need to verify signatures from a few hundred HTTPS domains with different certs for a given web request, and I'm not likely to open connections to a few hundred databases with different certs in a short amount of time, but the number of packages that are being discussed above are getting close to 500, presumably many of which are published by different entities and therefore signed with different certificates. That seems like a remarkably different scale from how X.509 certificates get used in those other circumstances, and I was concerned that the architectural choice itself to use X.509 compared to other types of signatures I've seen used by package managers (e.g. RSA keys, which I believe is what Arch used for its package signatures) might be responsible for the large overhead.

> I recently started using Arch Linux myself, and haven't had any problems with minute long restores.

This is super helpful information, and it probably is sufficient evidence to conclude that this is an issue on my end. I apologize for jumping to conclusions here; I made my previous comment in a moment of frustration, and in retrospect I should not have let that affect how I interacted here. I appreciate the patience in taking the time to respond to me despite that.

As for the suggestion for diagnosing this, `NUGET_CERT_REVOCATION_MODE=offline dotnet build` takes around 15-20 seconds to complete the restore step, which seems within the expected additional overhead from what the restore step takes with the verification disabled completely. At the risk of further diverging from the existing discussion on this issue, are there recommended steps I can take to verify what's going on with the revocation checks like you mention? I haven't noticed any issues with the network in the past that I used when running into this issue, and at least in terms of overall speed it's quite fast, but it seems plausible that it might be configured in some way that's causing this, and I'm not sure how I would approach trying to sort this out. (As mentioned above, I can open a separate issue if this is out of scope for debugging here).

**Comment by zivkan** [comment]
FWIW, I just had a go at restoring https://github.com/dotnet/orleans and https://github.com/OrchardCMS/OrchardCore.  In order to remove package source download speed as a variable, I'd restore once, copy the global packages folder to another location, and then edit the NuGet.config file to use the local directory as a package source instead.

Both would take between 4 to 5 seconds to restore, one downloaded 274 packages, the other 293, and using NUGET_CERT_REVOCATION_MODE=offline didn't seem to make any difference. The run to run variance was greater than the difference between online and offline for me, making it hard to determine if there was any difference.

It does make me wonder if I'm making a mistake, but I'm also not sure if openssl or something else is caching the certificate revocation information, or if it's disabled via some open ssl /etc config setting, because even if most of the packages are using the same certificates, there will still be at least a few dozen unique certificate chains to check and system monitor didn't show any meaningful network traffic during the restore.

But, I definitely can't replicate any problem where online revocation checks takes a minute or longer, while offline is only a few seconds.

> are there recommended steps I can take to verify what's going on with the revocation checks like you mention?

I don't know, I'm afraid.  I think NuGet just calls .NET's API to validate a certificate chain and then from my point of view it's up to the runtime, which then calls into openssl. We do have a bunch of certificate related code in NuGet.Packaging, but I don't think any of it checks revocation. But it's largely a guess, I can't remember what all that code is actually for.

When I ask an AI chatbot "does .NET have any etw events to investigate certificate validation performance issues" it says yes and gives some information. I guess it should be possible to use `dotnet-trace` to capture these EventSource events. Apparently PerfView can run on Linux, or maybe that's just the chatbot hallucinating. I only started using Linux at home as a daily driver less than a month ago, and not using it all during my work hours, so I'm not yet familiar with all the tooling used to investigate different types of issues.
